components/web_discovery/browser/patterns.h

```diff
@@ -0,0 +1,174 @@
+  // Coupled with the `kSingle` rule type.
+  kSingle,
+  // Coupled with the `kClustered` rule type.
+  kClustered,
+  // Currently unsupported/ignored.
+  kCustom
+};
+
+// Contains functions for refining the scraped value. The inner vector
+// contains the function name and arguments for the function.
+using RefineFunctionList = std::vector<base::Value::List>;
+
+// Defines rule for scraping an attribute from a given selected element.
+struct ScrapeRule {
+  ScrapeRule();
+  ~ScrapeRule();
+
+  ScrapeRule(const ScrapeRule&) = delete;
+  ScrapeRule& operator=(const ScrapeRule&) = delete;
+
+  // An optional selector for an element within the current selected element.
+  // The attribute will be retrieved from the embedded element.
+  std::optional<std::string> sub_selector;
+  ScrapeRuleType rule_type;
+  // The name of the attribute to retrieve for a DOM element.
+  std::string attribute;
+  // Functions used to refine the retrieved value. See the "func ids" defined
+  // in content_scraper.cc for all possible functions.
+  RefineFunctionList functions_applied;
+};
+
+// A map of keys (arbitrary IDs used for storing the scraped result) to scrape
+// rules.
+using ScrapeRuleGroup =
+    base::flat_map<std::string, std::unique_ptr<ScrapeRule>>;
+
+// A rule for provided a single key/value pair within the submission payload.
+struct PayloadRule {
+  PayloadRule();
+  ~PayloadRule();
+
+  PayloadRule(const PayloadRule&) = delete;
+  PayloadRule& operator=(const PayloadRule&) = delete;
+
+  // The DOM selector of the scraped attribute.
+  std::string selector;
+  // The arbitrary key associated with the scraped value.
+  std::string key;
+  // If set to true, an array-like Dict (each dict key is an index)
+  // will be rendered.
+  // Each value in the Dict will be a Dict containing all keys/values
+  // associated with the selector. This is commonly used for listing search
+  // results.
+  bool is_join = false;
+};
+
+// Contains rules for generating a payload for submission.
+struct PayloadRuleGroup {
+  PayloadRuleGroup();
+  ~PayloadRuleGroup();
+
+  PayloadRuleGroup(const PayloadRuleGroup&) = delete;
+  PayloadRuleGroup& operator=(const PayloadRuleGroup&) = delete;
+
+  // An arbitrary ID for the rule group. Current, this isn't used in the
+  // payload.
+  std::string key;
+  PayloadRuleType rule_type;
+  PayloadResultType result_type;
+  // The name of the "action" for the given payload.
+  std::string action;
+  // The rules for generating the fields within the payload.
+  std::vector<PayloadRule> rules;
+};
+
+// Contains settings and rule groups associated with a particular URL.
+struct PatternsURLDetails {
+  PatternsURLDetails();
+  ~PatternsURLDetails();
+
+  PatternsURLDetails(const PatternsURLDetails&) = delete;
+  PatternsURLDetails& operator=(const PatternsURLDetails&) = delete;
+
+  // The regex used to match the URL in the address bar.
+  std::unique_ptr<re2::RE2> url_regex;
+  bool is_search_engine;
```

Comment:
The fact that `is_search_engine` is not being initialised with any default value, and this type doesn't have a constructor to require a value here is being a cause for UB later down the line when this value is read by `IsPrivateURLLikely`. @DJAndries please advise how we should handle this, ideally by not having this value left not initialised.

Reply:
https://github.com/brave/brave-core/pull/29402